packages/jupyter-ai-magics/jupyter_ai_magics/partner_providers/copilot.py

```diff
@@ -0,0 +1,162 @@
+    "initializationOptions": {
+        "editorInfo": {"name": "JupyterLab", "version": jupyterlab_version},
+        "editorPluginInfo": {"name": "jupyter-ai", "version": jupyter_ai_version},
+    },
+}
+
+
+def calc_position_lineno_and_char(prefix, suffix):
+    """
+    Calculate the line number and character position within a text based on a given prefix and suffix text.
+    GitHub Copilot LSP requires those positions for completion requests.
+    https://www.npmjs.com/package/@github/copilot-language-server#panel-completions
+    """
+
+    full_text = prefix + suffix
+
+    lineno = full_text.count("\n", 0, len(prefix))
+    prefix_text = "\n".join(full_text.split("\n")[:lineno])
+    char_pos = len(prefix) - len(prefix_text) - 1
+
+    return lineno, char_pos
+
+
+class GitHubCopilotLLM:
+    process: Optional[Popen] = None
+
+    def __init__(self, lsp_bin_path: str):
+        self.lsp_bin_path = lsp_bin_path
+        self.ensure_lsp_server_initialized()
+
+    def _initialize(self):
+        self.lsp_endpoint.call_method(method_name="initialize", **INIT_PARAMS)
+        self.lsp_endpoint.send_notification(method_name="initialized")
+
+    def _signin(self):
+        self.ensure_lsp_server_initialized()
+        res = self.lsp_endpoint.call_method(
+            method_name="signIn",
+        )
+        return res
+
+    def _signout(self):
+        self.ensure_lsp_server_initialized()
+        res = self.lsp_endpoint.call_method(
+            method_name="signOut",
+        )
+        return res
+
+    def _completion(self, code, pos_line, pos_char):
+        self.ensure_lsp_server_initialized()
+        self.lsp_endpoint.send_notification(
+            method_name="textDocument/didOpen",
+            **{
+                "textDocument": {
+                    "uri": "file:///dummy",
+                    "version": 0,
+                    "languageId": "python",
+                    "text": code,
+                }
+            },
+        )
+
+        res = self.lsp_endpoint.call_method(
+            method_name="textDocument/copilotPanelCompletion",
+            **{
+                "textDocument": {
+                    "uri": "file:///dummy",
+                    "version": 0,
+                },
+                "position": {
+                    "line": pos_line,
+                    "character": pos_char,
+                },
+            },
+        )
+        return res
+
+    def _start_lsp_server(self):
+        if not self.is_lsp_server_running:
+            self.process = Popen(
+                [self.lsp_bin_path, "--stdio"], stdin=PIPE, stdout=PIPE, stderr=PIPE
+            )
+            self.json_rpc_endpoint = pylspclient.JsonRpcEndpoint(
+                self.process.stdin, self.process.stdout
+            )
+            self.lsp_endpoint = pylspclient.LspEndpoint(
+                self.json_rpc_endpoint, timeout=15
+            )
+            self.lsp_endpoint.start()
```

Comment:
Instead of adding a dependency on `pylspclient` it would probably be possible to use `jupyter-lsp` which is already a dependency of `jupyterlab` and `notebook`. I think the [`LanguageServerSession`](https://github.com/jupyter-lsp/jupyterlab-lsp/blob/73e390a8c53b30e46261315048da0cb5f3329385/python_packages/jupyter_lsp/jupyter_lsp/session.py#L23-L35) could be used here.